Add gemm solver in MIOpen

bghimireamd opened the pull request.

iq136boy committed:
[ci-skip] add invoke parameters header for gemm solver

iq136boy committed:
[ci-skip] add gemm problem description header

iq136boy committed:
[ci-skip] define solver header for gemm solver

iq136boy committed:
[ci-skip] add gemm headers to CMakeLists

iq136boy committed:
[ci-skip] add gemm header file that includes struct GemmNewDescriptor

iq136boy committed:
[ci-skip] add gemm cpp source file and commont out the gemm mode in header

iq136boy committed:
[ci-skip] add gemm header and gemm source files to CMakeLists

iq136boy committed:
[ci-skip] commont out miopenGemmDescriptor

iq136boy committed:
[ci-skip] minor fix

iq136boy committed:
[ci-skip] gemm new descriptor gemm probelm description

iq136boy committed:
[ci-skip] add gemm ocl source file

iq136boy committed:
[ci-skip] add MakeNetworkConfig() in gemm problem description source file

iq136boy committed:
[ci-skip] add definition of GetSolution in gemm_solver source file

iq136boy committed:
[ci-skip] add gemm solver source file in CMakeLists

iq136boy committed:
[ci-skip] bug fix

iq136boy committed:
[ci-skip] add leading dimensions, strides and batch count to gemm new descriptor

iq136boy committed:
[ci-skip] modified gemm invoke parameters (add A, B, C tensor descriptors)

iq136boy committed:
add gemm descriptor in miopen.h

iq136boy committed:
define miopenGemmmode_t

iq136boy committed:
define miopen gemm descriptor APIs

iq136boy committed:
add gemm api source file

iq136boy committed:
modified gemm apis,  gemm include header and gemm ocl

iq136boy committed:
modified the invoke parameters, ocl and header file

iq136boy committed:
[ci-skip] add gemm api source file to CMakelists

iq136boy committed:
[ci-skip] minor fix

iq136boy committed:
[ci-skip] commont out log functions in gemm api

iq136boy committed:
gemm network config function

iq136boy committed:
minor fix log cmd in gemm api

iq136boy committed:
gemm isapplicable and change solver name

iq136boy committed:
[ci-skip] solve the declaration warning by re-name the callgemm parameters

iq136boy committed:
minor fix

iq136boy committed:
add stream out operator overload for gemm solver

iq136boy committed:
uncomment MIOpen log function in gemm descriptor

iq136boy committed:
minor fix

iq136boy committed:
add gemm call rocblas header

iq136boy committed:
add gemm call rocblas cpp

iq136boy committed:
gemm rocblas source file to different place

iq136boy committed:
call gemm rocblas in gemm_solver get solution function

iq136boy committed:
simplify gemm rocblass code

bghimireamd committed:
bg/LWPMIOPEN-20: fixed merged issue

bghimireamd committed:
bg/LWPMIOPEN-20: fixed compilation issues

bghimireamd committed:
bg/LWPMIOPEN-20: complies but test is skippped, debuggind needed

bghimireamd committed:
bg/LWPMIOPEN-20: one example of CK gemm kernel test passed

bghimireamd committed:
bg/LWPMIOPEN-20 : add more tests

junliume committed:
fix clang-format issues

bghimireamd committed:
bg/LWPMIOPEN-20: add GEMM API test

bghimireamd committed:
bg/LWPMIOPEN-20: change name of ck gemm solver

bghimireamd committed:
bg/LWPMIOPEN-20 : clang-format

bghimireamd committed:
bg/LWPMIOPEN-20: fix merge conflict

bghimireamd committed:
bg/LWPMIOPEN-20: removed unnecessary files

bghimireamd committed:
bg/LWPMIOPEN-20 : fixed clang format

bghimireamd committed:
bg/LWPMIOPEN-20: skip test for GPU arch that CK does not support

bghimireamd committed:
bg/LWPMIOPEN-20 : make it ready to support for any data types

bghimireamd committed:
bg/LWPMIOPEN-20 : adding gemm + activ fusion, test is working

bghimireamd committed:
bg/LWPMIOPEN-20 : rename files to *gemm_activ*

bghimireamd committed:
bg/LWPMIOPEN-20 : use leading dimension

bghimireamd committed:
bg/LWPMIOPEN-20: remove unique id hack used for CK

bghimireamd committed:
bg/LWPMIOPEN-20 : add back all the test for gemm + activ

bghimireamd committed:
bg/LWPMIOPEN-20: clang format

bghimireamd committed:
bg/LWPMIOPEN-20: update commit has of CK in requirement.txt

bghimireamd committed:
bg/LWPMIOPEN-20 : fix ck commit

bghimireamd committed:
bg/LWPMIOPEN-20: fix test case

bghimireamd committed:
bg/LWPMIOPEN-20 : remove test with wrong dimensions

bghimireamd commented on driver/tensor_driver.hpp:
```diff
@@ -26,7 +26,8 @@
 
-#include "driver.hpp"
+//#include "driver.hpp"
+#define UNPACK_VEC4(v) (v[0]), (v[1]), (v[2]), (v[3])
```
You can ignore this for now. I have create another PR #1889 that address this issue.

bghimireamd commented on test/random.hpp:
```diff
@@ -41,7 +41,9 @@
 /// and vice versa. This independency could be important, because, for example, the driver
 /// implements its own cache of verification data and change or GET_RAND() would break it.
 
-static int GET_RAND()
```
You can ignore this for now. I have create another PR https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1889 that address this issue.

bghimireamd committed:
bg/LWPMIOPEN-20: remove unnecessarily large gemm test

bghimireamd committed:
bg/LWPMIOPEN-20 : fix merge conflict

bghimireamd committed:
bg/LWPMIOPEN-20: fix compile issue aftering merging with latest develop

bghimireamd committed:
bg/LWPMIOPEN-20 : remove unncessary code

bghimireamd committed:
bg/LWPMIOPEN-20: fix minor compile issue

bghimireamd committed:
bg/LWPMIOPEN-20 : fix compile issues

bghimireamd committed:
Merge branch 'develop' of github.com:ROCmSoftwarePlatform/MIOpen into bg/LWPMIOPEN-20

bghimireamd committed:
bg/LWPMIOPEN-20 : fix compile error after merging with develop

pfultz2 commented on include/miopen/miopen.h:
```diff
@@ -2690,6 +2699,27 @@ miopenDestroyActivationDescriptor(miopenActivationDescriptor_t activDesc);
+                                                      long long int ldA_,
+                                                      long long int ldB_,
+                                                      long long int ldC_,
+                                                      miopenDataType_t dataType_);
```
For convolution we dont include the input sizes and types in the convolution descriptor. We should not include it in the gemm descriptor either.

pfultz2 commented on src/fusion.cpp:
```diff
@@ -233,6 +233,21 @@ miopenStatus_t ConvForwardOpDescriptor::SetArgs(OperatorArgs& args,
 
+miopenStatus_t GemmOpDescriptor::GetOutputDesc(TensorDescriptor& output_desc) const
+{
+    std::vector<int> lens_in = {gemm_descriptor.GetM(), gemm_descriptor.GetN()};
```
The lens should be taken from the tensor descriptors.

pfultz2 commented on src/fusion.cpp:
```diff
@@ -233,6 +233,21 @@ miopenStatus_t ConvForwardOpDescriptor::SetArgs(OperatorArgs& args,
+miopenStatus_t GemmOpDescriptor::GetOutputDesc(TensorDescriptor& output_desc) const
+{
+    std::vector<int> lens_in = {gemm_descriptor.GetM(), gemm_descriptor.GetN()};
+    TensorDescriptor temp(input_desc.GetType(), input_desc.GetLayout_t(), lens_in);
```
The permutation(or layout) of the input matrices should not be propagated.

bghimireamd committed:
bg/LWPMIOPEN-20 : reduce test size to fix timeout issue during test case

bghimireamd committed:
bg/LWPMIOPEN-20 : remove timout from cmakelists

bghimireamd committed:
bg/LWPMIOPEN-20 : fixed merge conflict

bghimireamd committed:
bg/LWPMIOPEN-20 : check action in fusion plan

bghimireamd committed:
bg/LWPMIOPEN-20 : fix clang format

bghimireamd committed:
bg/LWPMIOPEN-20 : add FGELU activation in switch cases to avoid compilation errors

bghimireamd committed:
Merge branch 'develop' of github.com:ROCmSoftwarePlatform/MIOpen into bg/LWPMIOPEN-20

bghimireamd committed:
bg/LWPMIOPEN-20 : clang format

bghimireamd committed:
bg/LWPMIOPEN-20 : address review comments

bghimireamd committed:
bg/LWPMIOPEN-20 : fix null ptr issue

bghimireamd committed:
bg/LWPMIOPEN-20 : code cleanup

bghimireamd committed:
Merge branch 'develop' of github.com:ROCmSoftwarePlatform/MIOpen into bg/LWPMIOPEN-20

junliume commented:
@bghimireamd is this PR still WIP?

bghimireamd commented:
@junliume Yes, should be done next sprint.

JehandadKhan commented:
Closing this for now since @bghimireamd  is busy with other stuff.

DrizztDoUrden commented on src/CMakeLists.txt:
```diff
@@ -82,8 +82,10 @@ endfunction()
+    gemm/problem_description.cpp
     activ_api.cpp
     api/find2_0_commons.cpp
+    gemm_api.cpp
```
I'd move this to api/

DrizztDoUrden commented on src/CMakeLists.txt:
```diff
@@ -542,7 +546,7 @@ endif()
     list(APPEND MIOpen_Source
         gemm_v2.cpp
-        miopengemm.cpp
+        miopengemm.cpp      
```
Do we return miopengemm?

DrizztDoUrden commented on src/include/miopen/gemm/context.hpp:
```diff
@@ -0,0 +1,73 @@
+/// environmental context (e.g. HW/SW platform) and solver-specific state.
+///
+/// TODO: These three entities should be made separate.
+struct GemmContext : ExecutionContext
```
`ExecutionContext` shouldn't be extended. I have a PR to be merged soon, which among other things moves `is_for_generic_search` to `ExecutionContext`.

DrizztDoUrden commented on src/include/miopen/gemm/problem_description.hpp:
```diff
@@ -0,0 +1,123 @@
+    template <class Self>
+    static void Visit(Self&& self, std::function<void(int, std::string)> f)
+    {
+        // Once we tune gemm kernels we will not ignore.
```
This should either actually serialize or not implement `SQLiteSerializable`.
If it is expected, but not implemented yet, it should do `MIOPEN_THROW(miopenStatusNotImplemented)`.

DrizztDoUrden commented on src/ocl/fusionopbiasbnactivocl.cpp:
```diff
@@ -47,6 +47,13 @@ miopenStatus_t FusionOpDescriptor::GetNetworkConfig(std::stringstream& /*network
     return miopenStatusSuccess;
 }
 
+miopenStatus_t GemmOpDescriptor::GetNetworkConfig(std::stringstream& network_config,
```
This shouldn't take `Handle`.

DrizztDoUrden commented on src/solver.cpp:
```diff
@@ -292,6 +292,7 @@ inline SolverRegistrar::SolverRegistrar(IdRegistryData& registry)
              Primitive::Fusion,
              solver::fusion::ConvOclDirectFwdFused{}.SolverDbId(),
              miopenConvolutionAlgoDirect);
+    Register(registry, ++id, Primitive::Fusion, solver::fusion::CKGEMMActiv{}.SolverDbId());
```
This should only be added to the end of the file or it would corrupt existing serialized solutions.

atamazov commented:
@DrizztDoUrden This PR have been closed and branch deleted 2 months ago, any reason to review?

bghimireamd commented:
@atamazov I am currently tasked with integrating CK's GEMM fusion(GEMM + GELU) solver into MIOpen. This was an old PR of mine that already had done some grunt work, so wanted help from @DrizztDoUrden to see if what I have done so far is correct. We closed this PR last time since there was no proper infrastructure for find fusion solver. Since @DrizztDoUrden now have [find_fusion](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/2388) I will create another PR that will addressing all the issues mentioned here with [find_fusion](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/2388) integrated.

atamazov commented:
@bghimireamd Thanks for informing and for continuing this work!
